src/ril_voicecall.c

```diff
@@ -581,16 +581,37 @@ static void ril_voicecall_submit_hangup_req(struct ofono_voicecall *vc,
 {
 	struct ril_voicecall *vd = ril_voicecall_get_data(vc);
 	GRilIoRequest *ioreq = grilio_request_array_int32_new(1, id);
+	GSList *l;
+	struct ofono_call *call = NULL;
+	guint code;
 
 	/* Append the call id to the list of calls being released locally */
 	GASSERT(!gutil_int_array_contains(vd->local_release_ids, id));
 	gutil_int_array_append(vd->local_release_ids, id);
 
+	/* Look for call with id to get its status. */
+	for (l = vd->calls; l; l = l->next) {
+		call = l->data;
+		if (call->id == id)
+			break;
+		call = NULL;
+	}
+	/* Need to use this request so that declined
+	 * calls in this state, are properly forwarded
+	 * to voicemail.  REQUEST_HANGUP doesn't do the
+	 * right thing for some operators, causing the
+	 * caller to hear a fast busy signal or even not
+	 * hanging up the call, which rings again after
+	 * some seconds.
+	 */
+	code = call && call->status == OFONO_CALL_STATUS_INCOMING
+		? RIL_REQUEST_HANGUP_WAITING_OR_BACKGROUND : RIL_REQUEST_HANGUP;
+
 	/* Send request to RIL. ril_voicecall_request_data_free will unref
 	 * the request data */
 	req->ref_count++;
 	req->pending_call_count++;
-	grilio_queue_send_request_full(vd->q, ioreq, RIL_REQUEST_HANGUP,
+	grilio_queue_send_request_full(vd->q, ioreq, code,
```

Comment:
Unlike `RIL_REQUEST_HANGUP` (which takes connection index as a parameter), `RIL_REQUEST_HANGUP_WAITING_OR_BACKGROUND` takes no parameters.

In other words, `GRilIoRequest` has to be prepared differently for those requests.

Reply:
That's a good point, but I'm not sure where I can see this. Looking in `ofono-ril-binder-plugin/src/ril_binder_radio.c`, I found line 3382 that `RIL_REQUEST_HANGUP` has an encode function `ril_binder_radio_encode_ints` (I guess for the ids of the calls one want to hang-up, for instance in a conference call we are participating to), and `RIL_REQUEST_HANGUP_WAITING_OR_BACKGROUND` has an encode function `ril_binder_radio_encode_serial` which takes `req->current_id` (which is the default for every request, so meaning no parameter I presume). I guess the patch in its current form is working, because it is ignoring the int array in the background case, but that's not optimal ?

I'm pushing an additional commit with different preparations for the request. Tell me if it is the right way to do it. About checking the status of the background hangup, I'm still thinking how to handle this properly, everything being asynchronous...

Reply:
`grilio_request_array_int32_new(1, id)` allocates a request with one integer parameter, `grilio_request_new()` allocates a request without parameters.

Reply:
So, request initialisation looks good now, right ? I've squashed the commits so the PR is clean.